caom2pipe/caom2pipe/execute_composable.py

```diff
@@ -0,0 +1,1163 @@
+    with the checksum of the supposedly-the-same file stored at CADC.
+
+    This step should be invoked with any other task type that relies on
+    files on local disk.
+    """
+
+    def __init__(self, config, storage_name, command_name, cred_param,
+                 cadc_data_client, caom_repo_client):
+        super(Collection2CaomCompareChecksumClient, self).__init__(
+            config, mc.TaskType.CHECKSUM, storage_name, command_name,
+            cred_param, cadc_data_client, caom_repo_client)
+        self._define_local_dirs(storage_name)
+        self.fname = storage_name.get_fname_on_disk()
+
+    def execute(self, context):
+        self.logger.debug('Begin execute for {} '
+                          'CompareChecksum'.format(__name__))
+        self.logger.debug('the steps:')
+
+        self.logger.debug('generate the xml from the file on disk')
+        self._compare_checksums_client(self.fname)
+
+        self.logger.debug('End execute for {}'.format(__name__))
+
+
+class OrganizeExecutes(object):
+    """How to turn on/off various steps in the OMM pipeline."""
+    def __init__(self, config, todo_file=None):
+        self.config = config
+        self.task_types = config.task_types
+        self.logger = logging.getLogger()
+        self.logger.setLevel(config.logging_level)
+        if todo_file is not None:
+            self.todo_fqn = todo_file
+            todo_name = os.path.basename(todo_file).split('.')[0]
+            self.success_fqn = os.path.join(self.config.log_file_directory,
+                                            '{}_success.log'.format(todo_name))
+            self.failure_fqn = os.path.join(self.config.log_file_directory,
+                                            '{}_failure.log'.format(todo_name))
+            self.retry_fqn = os.path.join(self.config.log_file_directory,
+                                          '{}_retry.log'.format(todo_name))
+        else:
+            self.todo_fqn = config.work_fqn
+            self.success_fqn = config.success_fqn
+            self.failure_fqn = config.failure_fqn
+            self.retry_fqn = config.retry_fqn
+
+        if self.config.log_to_file:
+            mc.create_dir(self.config.log_file_directory)
+            failure = open(self.failure_fqn, 'w')
+            failure.close()
+            retry = open(self.retry_fqn, 'w')
+            retry.close()
+            success = open(self.success_fqn, 'w')
+            success.close()
+        self.success_count = 0
+        self.complete_record_count = 0
+
+    @property
+    def complete_record_count(self):
+        return self._complete_record_count
+
+    @complete_record_count.setter
+    def complete_record_count(self, value):
+        self._complete_record_count = value
+
+    def choose(self, storage_name, command_name, meta_visitors=None,
+               data_visitors=None):
+        executors = []
+        if storage_name.is_valid():
```

Comment:
I agree that this method is complicated. However, if I were to handle the logic within the classes themselves, the number of locations to edit would increase quickly. For example, for the "use_local_files" flag, I would have to edit five execute methods to have the check in there (one for each of SCRAPE, INGEST, MODIFY, STORE, and VISIT task types), instead of this one method. It's a different way of arranging the complexity, hopefully localizing it to this one method, instead of across many execute methods.

I don't know that it will prove out, but so far it's been pretty nice in terms of adding new flags and new functionality.